routing/payment_lifecycle.go

```diff
@@ -203,20 +203,41 @@ func (p *paymentLifecycle) resumePayment(ctx context.Context) ([32]byte,
 		return [32]byte{}, nil, err
 	}
 
-	// Get the payment status.
-	status := payment.GetStatus()
-
 	// exitWithErr is a helper closure that logs and returns an error.
-	exitWithErr := func(err error) ([32]byte, *route.Route, error) {
-		// Log an error with the latest payment status.
-		//
-		// NOTE: this `status` variable is reassigned in the loop
-		// below. We could also call `payment.GetStatus` here, but in a
-		// rare case when the critical log is triggered when using
-		// postgres as db backend, the `payment` could be nil, causing
-		// the payment fetching to return an error.
-		log.Errorf("Payment %v with status=%v failed: %v", p.identifier,
-			status, err)
+	exitWithErr := func(exitErr error) ([32]byte, *route.Route, error) {
+		// We fetch the latest payment status here to get the latest
+		// state of the payment.
+		payment, err := p.router.cfg.Control.FetchPayment(p.identifier)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		// Only fail the payment if there are no inflight HTLCs on it
+		// because otherwise the payment might be resolved when
+		// resuming.
+		_, failure := payment.TerminalInfo()
+		if failure == nil &&
```

Comment:
I don't think we should make it this complicated - if we really want to fail the payment, I think the only safe case to fail is when the status is in `StatusInitiated`, which is the case from using `keysend` + `max_shard_size_sat`, the payment was never attempted.